mentat/code_context.py

```diff
@@ -199,147 +117,306 @@ async def refresh(self, replace_paths: bool | None = False):
+                path = f"{path}:{interval_str}"
+            diff_target = (
+                self.diff_context.target if path in self.diff_context.files else None
+            )
+            feature = CodeFile(path, feature.level, diff=diff_target)
+            include_features.append(feature)
+
+        def _feature_relative_path(f: CodeFile) -> str:
+            return os.path.relpath(f.path, self.config.git_root)
+
+        return sorted(include_features, key=_feature_relative_path)
+
+    async def _get_auto_features(
+        self,
+        code_file_manager: "CodeFileManager",
+        parser: "Parser",
+        model: str,
+        include_features: list[CodeFile],
+        max_tokens: Optional[int] = None,
+    ) -> list[CodeFile]:
+        # Generate all possible permutations for all files in the project.
+        candidate_features = list[CodeFile]()
+        for path in get_non_gitignored_files(self.config.git_root):
+            if not path.is_dir() and not is_file_text_encoded(path):
+                continue
+            permutations: list[tuple[CodeMessageLevel, str | None]] = [
+                (CodeMessageLevel.CODE, None),
+                (CodeMessageLevel.FILE_NAME, None),
+            ]
+            if not self.settings.no_code_map:
+                permutations += [
+                    (CodeMessageLevel.CMAP_FULL, None),
+                    (CodeMessageLevel.CMAP, None),
+                ]
+            if self.diff_context.target and path in self.diff_context.files:
+                permutations += [
+                    (level, self.diff_context.target) for level, _ in permutations
                 ]
-                message = "\n".join(message)
-                await stream.send(message, color="yellow")
+            for level, diff in permutations:
+                feature = CodeFile(path, level=level, diff=diff)
+                candidate_features.append(feature)
 
-        return "\n".join(code_message)
+        # Sort candidates by relevance/density.
+        tokens_remaining = 1e6 if max_tokens is None else max_tokens
 
-    async def add_file(self, code_file: CodeFile):
-        stream = SESSION_STREAM.get()
+        async def _calculate_feature_score(feature: CodeFile) -> float:
+            score = 0.0
+            tokens = await feature.count_tokens(
+                self.config, code_file_manager, parser, model
+            )
+            if tokens == 0:
+                raise MentatError(f"Feature {feature} has 0 tokens.")
+            if feature.diff is not None:
+                score += 1
+            if feature.level == CodeMessageLevel.FILE_NAME:
+                score += 0.1
+            if feature.level == CodeMessageLevel.CMAP:
+                score += 0.25
+            elif feature.level == CodeMessageLevel.CMAP_FULL:
+                score += 0.5
+            elif feature.level == CodeMessageLevel.CODE:
+                score += 0.75
```

Comment:
I ran this on the mentat repo, here's how it looks:
![image](https://github.com/AbanteAI/mentat/assets/50287275/a5e33fc7-8cda-4c9e-b161-4a115286161d)
![image](https://github.com/AbanteAI/mentat/assets/50287275/4b058fb3-bc59-467d-b330-c20a51dd04aa)

Reply:
So if I'm understanding correctly this algorithm really wants to include the cmap and really doesn't want to include the full file? I suppose that makes sense since the cmap will be far fewer tokens and is only weighted half.

Reply:
> So if I'm understanding correctly this algorithm really wants to include the cmap and really doesn't want to include the full file? I suppose that makes sense since the cmap will be far fewer tokens and is only weighted half.

Yeah, right now we don't have a way to prioritize, so this is pretty similar to what we had before, but more granular (before we just added the biggest cmap level that could fit for the whole codebase, now we'll do different levels in different places).

Once we have some embeddings to prioritize code we'll need to make sure the balance between cmaps and full code makes sense. Would be silly to have the algorithm be super biased all the way in either direction